protocol/common/message_segmentation.go

```diff
@@ -0,0 +1,361 @@
+	if err != nil {
+		return err
+	}
+
+	if len(segments) == 0 {
+		return errors.New("unexpected state: no segments found after save operation") // This should theoretically never occur.
+	}
+
+	firstSegmentMessage := segments[0]
+	lastSegmentMessage := segments[len(segments)-1]
+
+	// First segment message must not be a parity message.
+	if firstSegmentMessage.IsParityMessage() || len(segments) != int(firstSegmentMessage.SegmentsCount) {
+		return ErrMessageSegmentsIncomplete
+	}
+
+	payloads := make([][]byte, firstSegmentMessage.SegmentsCount+lastSegmentMessage.ParitySegmentsCount)
+	payloadSize := len(firstSegmentMessage.Payload)
+
+	restoreUsingParityData := lastSegmentMessage.IsParityMessage()
+	if restoreUsingParityData {
+		enc, err := reedsolomon.New(int(firstSegmentMessage.SegmentsCount), int(lastSegmentMessage.ParitySegmentsCount))
+		if err != nil {
+			return err
+		}
+
+		var lastNotParitySegmentPayload []byte
+		for _, segment := range segments {
+			if !segment.IsParityMessage() {
+				if segment.Index == firstSegmentMessage.SegmentsCount-1 {
+					// Ensure last segment is aligned to payload size, as it is required by reedsolomon.
+					payloads[segment.Index] = make([]byte, payloadSize)
+					copy(payloads[segment.Index], segment.Payload)
+					lastNotParitySegmentPayload = segment.Payload
+				} else {
+					payloads[segment.Index] = segment.Payload
+				}
+			} else {
+				payloads[firstSegmentMessage.SegmentsCount+segment.ParitySegmentIndex] = segment.Payload
+			}
+		}
+
+		err = enc.Reconstruct(payloads)
+		if err != nil {
+			return err
+		}
+
+		ok, err := enc.Verify(payloads)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return ErrMessageSegmentsInvalidParity
+		}
+
+		if lastNotParitySegmentPayload != nil {
+			payloads[firstSegmentMessage.SegmentsCount-1] = lastNotParitySegmentPayload // Bring back last segment with original length
+		}
+	} else {
+		for i, segment := range segments {
+			payloads[i] = segment.Payload
+		}
+	}
+
+	// Combine payload
+	var entirePayload bytes.Buffer
+	for i := 0; i < int(firstSegmentMessage.SegmentsCount); i++ {
+		_, err := entirePayload.Write(payloads[i])
+		if err != nil {
+			return errors.Wrap(err, "failed to write segment payload")
+		}
+	}
+
+	// Sanity check
+	entirePayloadHash := crypto.Keccak256(entirePayload.Bytes())
+	if !bytes.Equal(entirePayloadHash, segmentMessage.EntireMessageHash) {
+		return ErrMessageSegmentsHashMismatch
+	}
+
+	err = s.persistence.CompleteMessageSegments(segmentMessage.EntireMessageHash, message.TransportLayer.SigPubKey, time.Now().Unix())
+	if err != nil {
+		return err
+	}
+
+	message.TransportLayer.Payload = entirePayload.Bytes()
+
+	return nil
+}
+
+func (s *MessageSender) CleanupSegments() error {
+	weekAgo := time.Now().AddDate(0, 0, -7).Unix()
+	monthAgo := time.Now().AddDate(0, -1, 0).Unix()
+
+	err := s.persistence.RemoveMessageSegmentsOlderThan(weekAgo)
```

Comment:
Hmmm, but we could fetch the missing segments later from store nodes, couldn't we?

Reply:
Store nodes store up to 30 days, but we fetch only 7-9 days, so it works. But Maybe using a constant used in both would make more sense.

Also, does this clean up work on all segments or only the orphan ones? ie when a message is completed (we got all needed segments), are the segments kept 7 days or are they cleaned up automatically after?

edit: I just saw the `RemoveMessageSegmentsCompletedOlderThan(monthAgo)`. Shouldn't we clean up completed segments as soon as we complete why do we keep them?

Reply:
> Hmmm, but we could fetch the missing segments later from store nodes, couldn't we?

That's right. I assumed that if one couldn't fetch the full message for a week, then the missing segment would never come. However, I overlooked the case when one goes on a two-week vacation and then fetches from store nodes. I changed both cleanups to a month interval.

> Also, does this clean up work on all segments or only the orphan ones? ie when a message is completed (we got all needed segments), are the segments kept 7 days or are they cleaned up automatically after?

They are cleaned up automatically after, see: `CompleteMessageSegments`.

> edit: I just saw the RemoveMessageSegmentsCompletedOlderThan(monthAgo). Shouldn't we clean up completed segments as soon as we complete why do we keep them?

The name is a bit unfortunate. It cleans up only the table with hashes, not the message segments themselves. There is a table named `message_segments_completed` that holds the hash of the completed message and the `timestamp` of when it was completed. This is to ignore segments of messages that were already completed.